x-pack/plugins/cases/server/services/user_actions/transform.ts

```diff
@@ -0,0 +1,316 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+/* eslint-disable @typescript-eslint/naming-convention */
+
+import * as rt from 'io-ts';
+import { isString } from 'lodash';
+
+import { SavedObjectReference } from '../../../../../../src/core/server';
+import {
+  CaseAttributes,
+  CaseConnector,
+  CaseConnectorRt,
+  CaseExternalServiceBasicRt,
+  isCreateConnector,
+  isPush,
+  isUpdateConnector,
+  noneConnectorId,
+} from '../../../common';
+import {
+  CONNECTOR_ID_REFERENCE_NAME,
+  getNoneCaseConnector,
+  PUSH_CONNECTOR_ID_REFERENCE_NAME,
+  USER_ACTION_OLD_ID_REF_NAME,
+  USER_ACTION_OLD_PUSH_ID_REF_NAME,
+} from '../../common';
+import { ACTION_SAVED_OBJECT_TYPE } from '../../../../actions/server';
+import { UserActionFieldType } from './types';
+
+export function extractConnectorIdFromJson({
+  action,
+  actionFields,
+  actionDetails,
+  fieldType,
+}: {
+  action?: string;
+  actionFields?: string[];
+  actionDetails?: string | null;
+  fieldType: UserActionFieldType;
+}): { transformedActionDetails?: string | null; references: SavedObjectReference[] } {
+  const defResult = { transformedActionDetails: actionDetails, references: [] };
+
+  if (!action || !actionFields || !actionDetails) {
+    return defResult;
+  }
+
+  try {
+    const decodedJson = JSON.parse(actionDetails);
+
+    const { transformedActionDetails, references } = extractConnectorIdHelper({
+      action,
+      actionFields,
+      actionDetails: decodedJson,
+      fieldType,
+    });
+
+    return { transformedActionDetails, references };
+  } catch (error) {
+    return defResult;
+  }
+}
+
+export function extractConnectorId({
+  action,
+  actionFields,
+  actionDetails,
+  fieldType,
+}: {
+  action: string;
+  actionFields: string[];
+  actionDetails?: Record<string, unknown> | string | null;
+  fieldType: UserActionFieldType;
+}): {
+  transformedActionDetails?: string | null;
+  references: SavedObjectReference[];
+} {
+  if (!actionDetails || isString(actionDetails)) {
+    // the action was null, undefined, or a regular string so just return it unmodified and not encoded
+    return { transformedActionDetails: actionDetails, references: [] };
+  }
+
+  try {
+    return extractConnectorIdHelper({
```

Comment:
This handles encoding the actionDetails. If that failed for some reason we'll return whatever was passed to use and whatever was passed to use and convert it to a string.